File: app/src/main/java/omniex/nl/omniex/ui/app/main/MainMenuPresenter.java
package omniex.nl.omniex.ui.app.main;

import com.google.gson.JsonSyntaxException;

import javax.inject.Inject;

import nl.omniex.omniexshopping.data.model.cart.CartItemDelete;
import nl.omniex.omniexshopping.data.model.cart.CartQuantitySetter;
import nl.omniex.omniexshopping.data.repository.CartRepository;
import nl.omniex.omniexshopping.data.repository.OrderRepository;
import nl.omniex.omniexshopping.data.repository.ProfileRepository;
import nl.omniex.omniexshopping.ui.base.BasePresenter;
import nl.omniex.omniexshopping.ui.base.BaseView;
import nl.omniex.omniexshopping.utils.SharedPrefUtils;
import timber.log.Timber;

public class MainMenuPresenter extends BasePresenter<MainMenuView> {

    private ProfileRepository mProfileRepository;
    private CartRepository mCartRepository;
    private OrderRepository mOrderRepository;

    @Inject
    MainMenuPresenter(ProfileRepository profileRepository, CartRepository cartRepository, OrderRepository orderRepository) {
        mProfileRepository = profileRepository;
        mCartRepository = cartRepository;
        mOrderRepository = orderRepository;
    }

    void logout() {
        addDisposable(mProfileRepository
                .logout()
                .subscribe(
                        voidResponse -> {
                            if (voidResponse.code() == 200) {
                                SharedPrefUtils.setUserLogged(false);
                                ifViewAttached(MainMenuView::onLogoutSuccess);
                            }
                        }, Throwable::printStackTrace));
    }

    void getCart() {
        ifViewAttached(BaseView::startLoading);
        addDisposable(mCartRepository
                .getCart()
                .subscribe(
                        cartResponse ->
                        {
                            ifViewAttached(view -> {
                                view.onCartFetched(cartResponse.body().getCart());
                                view.stopLoading();
                            });
                        },
                        error -> {
                            error.printStackTrace();
                            if(error instanceof JsonSyntaxException)
                                ifViewAttached(view -> view.onCartEmpty());
                            Timber.e(error);
                            ifViewAttached(BaseView::stopLoading);
                        }));
    }

    void updateCartQuantity(CartQuantitySetter cartQuantitySetter) {
        addDisposable(mCartRepository
                .updateCartItemQuantity(cartQuantitySetter)
                .subscribe(voidResponse -> {
                    if(voidResponse.code()==200)
                        ifViewAttached(MainMenuView::onCartItemQuantityUpdated);
                }, Throwable::printStackTrace));
    }

    void deleteCartItem(CartItemDelete cartItemDelete) {
        addDisposable(mCartRepository
                .deleteCartItem(cartItemDelete)
                .subscribe(voidResponse -> {
                    if(voidResponse.code()==200)
                        ifViewAttached(MainMenuView::onCartItemQuantityUpdated);
                }, Throwable::printStackTrace));
    }
}

File: app/src/main/java/omniex/nl/omniex/ui/app/main/MainMenuPresenter.kt
package omniex.nl.omniex.ui.app.main

import com.google.gson.JsonSyntaxException
import omniex.nl.omniex.data.model.cart.CartItemDelete
import omniex.nl.omniex.data.model.cart.CartQuantitySetter
import omniex.nl.omniex.data.repository.CartRepository
import omniex.nl.omniex.data.repository.OrderRepository
import omniex.nl.omniex.data.repository.ProfileRepository
import omniex.nl.omniex.ui.base.BasePresenter
import omniex.nl.omniex.utils.SharedPrefUtils
import timber.log.Timber
import javax.inject.Inject

class MainMenuPresenter @Inject
internal constructor(private val mProfileRepository: ProfileRepository, private val mCartRepository: CartRepository, private val mOrderRepository: OrderRepository) : BasePresenter<MainMenuView>() {

    fun logout() {
        addDisposable(mProfileRepository
                .logout()
                .subscribe(
                        { voidResponse ->
                            if (voidResponse.code() === 200) {
                                SharedPrefUtils.setUserLogged(false)
                                ifViewAttached { it.onLogoutSuccess() }
                            }
                        }, {
                    it.printStackTrace()
                }))
    }

    fun getCart() {
        ifViewAttached { it.startLoading() }
        addDisposable(mCartRepository
                .cart
                .subscribe(
                        { cartResponse ->
                            ifViewAttached { view ->
                                view.onCartFetched(cartResponse.body()!!.cart!!)
                                view.stopLoading()
                            }
                        },
                        { error ->
                            error.printStackTrace()
                            if (error is JsonSyntaxException)
                                ifViewAttached { view -> view.onCartEmpty() }
                            Timber.e(error)
                            ifViewAttached { it.stopLoading() }
                        }))
    }

    fun updateCartQuantity(cartQuantitySetter: CartQuantitySetter) {
        addDisposable(mCartRepository
                .updateCartItemQuantity(cartQuantitySetter)
                .subscribe({ voidResponse ->
                    if (voidResponse.code() === 200)
                        ifViewAttached { it.onCartItemQuantityUpdated() }
                }, { it.printStackTrace() }))
    }

    fun deleteCartItem(cartItemDelete: CartItemDelete) {
        addDisposable(mCartRepository
                .deleteCartItem(cartItemDelete)
                .subscribe({ voidResponse ->
                    if (voidResponse.code() === 200)
                        ifViewAttached { it.onCartItemQuantityUpdated() }
                }, { it.printStackTrace() }))
    }
}
